### PR TITLE
[IMP] websale_stock: add email notification

### PR DESCRIPTION
**Pull request by alpa-odoo**
This PR adds a message "Be notified when in stock" on a product when it is out of stock. When clicking the message, a form appears where the user can enter his email. If he is a logged user, the email associated to the partner of the user is pre-filled in the form. Otherwise the user enters his email. Afterwards, the user clicks on the button and the following things take place:

1. A `product_wishlist` is created with the field `stock_notification` to True with a partner associated with the email.
2. Afterwards, the cron in `addons/website_sale_stock/data/ir_cron_data.xml` will run and send an email to every wishlist email for which the `stock_notification` is True

From a technical point of view, it moves the model `product_wishlist` from: 
`website_sale_wishlist` -> `website_sale`.

Since `website_sale_stock` will use this model to we decided that it was better to put it there and inherit it in `website_sale_wishlist` and `website_sale_stock`

task-id 2806828

### CODE REVIEW

File: odoo/service/server.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def set_limit_memory_hard():
     if os.name == 'posix' and config['limit_memory_hard']:
         rlimit = resource.RLIMIT_RSS if platform.system() == 'Darwin' else resource.RLIMIT_AS
         soft, hard = resource.getrlimit(rlimit)
-        resource.setrlimit(rlimit, (config['limit_memory_hard'], hard))
+        # resource.setrlimit(rlimit, (config['limit_memory_hard'], hard))
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/service/server.py:83]: instead of commenting the limit, start odoo-bin with `--limit-memory-hard 9999999999`

[the remark this answers is not in the document]
Comment (alpa-odoo) [reply odoo/service/server.py:83]: Thanks !!!

[the remark this answers is not in the document]
Comment (alpa-odoo) [reply odoo/service/server.py:83]: Screenshot 2022-04-01 at 17 05 19

File: addons/website_sale_stock/static/src/js/variant_mixin.js [hunk 2]
```diff
@@ -82,6 +82,15 @@ VariantMixin._onChangeCombinationStock = function (ev, $parent, combination) {
 };
 
 publicWidget.registry.WebsiteSale.include({
+
+    events: Object.assign(publicWidget.registry.WebsiteSale.prototype.events, {
+         'click .show_notification_email_form': '_onClickShowNotificationEmailForm'
+    }),
```
Comment (Williambraecky) [inline addons/website_sale_stock/static/src/js/variant_mixin.js:88]: This is modifying in place and assigning events at the same time
```suggestion
    events: Object.assign({}, publicWidget.registry.WebsiteSale.prototype.events, {
         'click .show_notification_email_form': '_onClickShowNotificationEmailForm'
    }),
```

File: addons/website_sale_stock/controllers/main.py [hunk 3]
```diff
@@ -30,3 +32,38 @@ def shop_payment_transaction(self, *args, **kwargs):
...
+            partner_id = request.env.user.partner_id
+
+        wish = wishlist_model._add_to_wishlist(
+            pricelist.id,
+            pricelist.currency_id.id,
+            request.website.id,
+            price,
+            product_id,
+            partner_id
+        )
```
Comment (Williambraecky) [inline addons/website_sale_stock/controllers/main.py:66]: To check with the PO: this will allow a public user to register an existing user to products. We might need to check that the partner/mail isn't linked to an existing user as well.

File: addons/website_sale/models/product_wishlist.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    def _add_to_wishlist(self, pricelist_id, currency_id, website_id, price, product_id, partner_id=False):
+        wish = self.env['product.wishlist'].create({
+            'partner_id': partner_id,
+            'product_id': product_id,
+            'currency_id': currency_id,
+            'pricelist_id': pricelist_id,
+            'price': price,
+            'website_id': website_id,
+        })
+        return wish
```
Comment (Williambraecky) [inline addons/website_sale/models/product_wishlist.py:35]: Since the route calling this method is public, we also need to check that the wishlist does not exist, we would otherwise trigger the unicity rules defined above.
```suggestion
    def _add_to_wishlist(self, pricelist_id, currency_id, website_id, price, product_id, partner_id=False):
        wish = self.env['product.wishlist'].sudo().search([('partner_id', '=', partner_id), ('product_id', '=', product_id)]) # Unsure about sudo here
        if wish:
            # Maybe update price pricelist etc
        else:
            wish = self.env['product.wishlist'].create({
                'partner_id': partner_id,
                'product_id': product_id,
                'currency_id': currency_id,
                'pricelist_id': pricelist_id,
                'price': price,
                'website_id': website_id,
            })
        return wish
```

File: addons/website_sale/models/product_wishlist.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+from datetime import datetime, timedelta
```
Comment (Williambraecky) [inline addons/website_sale/models/product_wishlist.py:1]: Missing license info
```suggestion
# -*- coding: utf-8 -*-
# Part of Odoo. See LICENSE file for full copyright and licensing details.

from datetime import datetime, timedelta
```

File: addons/website_sale_stock/static/src/js/variant_mixin.js [hunk 6]
```diff
@@ -82,6 +82,15 @@ VariantMixin._onChangeCombinationStock = function (ev, $parent, combination) {
...
+    _onClickShowNotificationEmailForm: function (ev) {
+        //hide the message, display the input box
+        $(ev.currentTarget).hide();
+        $('.email_notification_form').removeClass('d-none');
```
Comment (Williambraecky) [inline addons/website_sale_stock/static/src/js/variant_mixin.js:92]: If possible, try not to use jQuery

File: addons/website_sale_stock/static/src/xml/website_sale_stock_product_availability.xml [hunk 7]
```diff
@@ -8,6 +8,20 @@
...
+                        <div class="input-group w-75">
+                            <input name="email" class="form-control" type="text" placeholder="john.doe@example.com"/>
+                            <input type="hidden" name="product_id" t-att-value="product_id"/>
+                            <div class="input-group-append">
+                                <a href="#" role="button" class="btn btn-secondary a-submit">Get Notified</a>
+                            </div>
+                        </div>
+                    </form>
+                </div>
+            </t>
```
Comment (Williambraecky) [inline addons/website_sale_stock/static/src/xml/website_sale_stock_product_availability.xml:24]: This has the exact same condition as the div (`id="out_of_stock_message"`) above, can't it be part of it?

File: addons/website_sale_stock/views/website_sale_stock_templates.xml [hunk 8]
```diff
@@ -3,51 +3,60 @@
...
+                <div class="input-group-append">
+                    <a t-attf-href="#" class="btn btn-link">
+                        <i class='fa fa-warning text-warning' t-att-title="line._get_stock_warning()" role="img"
+                           aria-label="Warning"/>
+                    </a>
+                </div>
+            </t>
+            <t t-else="1">
+                <t>$0</t>
+            </t>
```
Comment (Williambraecky) [inline addons/website_sale_stock/views/website_sale_stock_templates.xml:22]: Avoid linting automatically :)

File: addons/website_sale_stock/views/website_sale_stock_templates.xml [hunk 9]
```diff
@@ -3,51 +3,60 @@
...
+        </xpath>
+        <xpath expr="//table[@id='cart_products']" position="after">
+            <t t-if='website_sale_order'>
+                <t t-set='warning' t-value='website_sale_order._get_stock_warning(clear=False)'/>
+                <div t-if='warning' class="alert alert-warning" role="alert">
+                    <strong>Warning!</strong>
+                    <t t-esc='website_sale_order._get_stock_warning()'/>
+                </div>
+            </t>
```
Comment (Williambraecky) [inline addons/website_sale_stock/views/website_sale_stock_templates.xml:57]: This seems to also be just linting?

File: addons/website_sale_stock_wishlist/models/__init__.py [hunk 10]
```diff
@@ -1,5 +1,4 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from . import product_wishlist
```
Comment (Williambraecky) [inline addons/website_sale_stock_wishlist/models/__init__.py:4]: You probably need to rever this change.

**Review by Williambraecky — COMMENTED** [review]
Smol review

File: addons/website_sale/models/product_wishlist.py (hunk 5 above)
Comment (Julien00859) [reply addons/website_sale/models/product_wishlist.py:1]: coding utf-8 is useless since py3

File: addons/website_sale_stock_wishlist/models/__init__.py (hunk 10 above)
Comment (alpa-odoo) [reply addons/website_sale_stock_wishlist/models/__init__.py:4]: Even if product_wishlist.py is not in website_sale_stock_wishlist anymore 
Screenshot 2022-04-05 at 14 17 09
?

File: addons/website_sale_stock_wishlist/static/src/js/website_sale.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            return
+        }
+        this._rpc({
+            route: "/shop/add_stock_email_notification",
+            params: {
+                product_id,
+                email
+            },
+        }).then(function (data) {
+            if (data === '400 Bad Request: Invalid Email'){
```
Comment (Williambraecky) [inline addons/website_sale_stock_wishlist/static/src/js/website_sale.js:37]: I am not a fan of this condition.
I think we never use `data`. Could we check if `data` is a string instead and consider that if `data` is a string, it's an error?

File: addons/website_sale_wishlist/static/src/js/website_sale.js [hunk 12]
```diff
@@ -1,8 +1,17 @@
-/** @odoo-module **/
+/** @odoo-module alias=website_sale_wishlist.website_sale **/
 
-import { WebsiteSale } from 'website_sale.website_sale';
+import {WebsiteSale} from 'website_sale.website_sale';
```
Comment (Williambraecky) [inline addons/website_sale_wishlist/static/src/js/website_sale.js:3]:
```suggestion
import { WebsiteSale } from 'website_sale.website_sale';
```

File: addons/website_sale_wishlist/static/src/js/website_sale.js [hunk 13]
```diff
@@ -13,4 +22,276 @@ WebsiteSale.include({
...
+            route: "/shop/cart/update_json", params: this._getCartUpdateJsonParams(productID, qty),
+        }).then(function (resp) {
+            if (resp.warning) {
+                if (!$('#data_warning').length) {
+                    $('.wishlist-section').prepend('<div class="mt16 alert alert-danger alert-dismissable" role="alert" id="data_warning"></div>');
+                }
+                let cart_alert = $('.wishlist-section').parent().find('#data_warning');
+                cart_alert.html('<button type="button" class="close" data-dismiss="alert" aria-hidden="true">&times;</button> ' + resp.warning);
+            }
+            $('.my_cart_quantity').html(resp.cart_quantity || '<i class="fa fa-warning" /> ');
```
Comment (Williambraecky) [inline addons/website_sale_wishlist/static/src/js/website_sale.js:193]: For security:
The code was moved, it already existed in the following file: `addons/website_sale_wishlist/static/src/js/website_sale_wishlist.js`

File: addons/website_sale_stock/data/template_email.xml [hunk 14]
```diff
@@ -11,6 +11,12 @@
             </div>
             <div style="display: flex; flex-direction: row; align-items: center; justify-content: center; width: 100%;">
                 <p t-esc="wishlist.product_id.name"/>
+                <p t-if="wishlist.product_id.product_template_attribute_value_ids"
+                   style="margin-left: 0.5em;">
+                    (<p
+                    t-out='wishlist.product_id.product_template_attribute_value_ids.name'
+                    />)
+                </p>
```
Comment (Williambraecky) [inline addons/website_sale_stock/data/template_email.xml:19]: Why this change by the way?

File: addons/website_sale_stock/models/product_wishlist.py [hunk 15]
```diff
@@ -33,15 +16,19 @@ def _send_availability_email(self):
...
+                                                        {"wishlist": wishlist})
                 msg = self.env["mail.message"].sudo().new(dict(body=body_html, record_name=product.name))
                 full_mail = self.env["mail.render.mixin"]._render_encapsulate(
                     "mail.mail_notification_light",
                     body_html,
                     add_context=dict(message=msg, model_description=_("Wishlist")),
                 )
+                product_name = product.name
+                if product.product_template_attribute_value_ids:
+                    product_name += f" ({product.product_template_attribute_value_ids.name})"
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/website_sale_stock/models/product_wishlist.py:29]: Same here, why this change?

**Review by Williambraecky — CHANGES_REQUESTED** [review]
Thanks for your work 🙂  good job.
The pr got quite big 🤯

File: addons/website_sale_stock/models/product_wishlist.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (alpa-odoo) [reply addons/website_sale_stock/models/product_wishlist.py:29]: MJU had the nice idea to make the mail template reflect the fact that the notification was on a given product variant:

Screenshot 2022-05-30 at 16 19 03

https://watch.screencastify.com/v/vTyYgopngR23ck521q9R


So I followed his feedback and modified the code to reflect it:

Screenshot 2022-05-30 at 16 22 00

File: addons/website_sale_stock/static/src/js/website_sale.js [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+                return
+            }
+            const div = form.closest('#add_stock_email_notification_product_div');
+            const message = div.querySelector('#add_stock_email_notification_product_success_message');
+
+            message.classList.remove('d-none');
+            form.classList.add('d-none');
+
+            // This is to add the product to the wishlist since it creates one.
+            // We have to check that website_sale_wishlist is installed to add it.
```
[the remark this answers is not in the document]
Comment (alpa-odoo) [reply addons/website_sale_stock/static/src/js/website_sale.js:55]: Screenshot 2022-05-31 at 10 44 54

This is to trigger the animation that adds a product in the wishlist so I mean that `website_stock` could be installed and not `website_sale_wishlist`, in which case `self.wishlistProductIDs` wouldn't exists since it is added by `website_sale_wishlist`.

Another idea would be to do that in website_sale_stock_wishlist by adding behaviour to the `_onSubmitAddStockEmailNotificationProductForm` function since website_sale_stock_wishlist depends on `website_sale_wishlist` and `website_sale_wishlist` ? What do you think ?

File: addons/website_sale_stock_wishlist/static/src/js/website_sale.js (hunk 11 above)
Comment (alpa-odoo) [reply addons/website_sale_stock_wishlist/static/src/js/website_sale.js:37]: Screenshot 2022-05-31 at 12 54 51

The thing is that here, there is no error, the server returns a string which contains `"product.wishlist(52,)"`. Is there a way to check the status code ?

File: addons/website_sale_stock_wishlist/static/src/js/variant_mixin.js [hunk 17]
```diff
@@ -1,7 +1,6 @@
 /** @odoo-module **/
 
 import VariantMixin from "website_sale_stock.VariantMixin";
-import "website_sale.website_sale";
```
[the remark this answers is not in the document]
Comment (alpa-odoo) [reply addons/website_sale_stock_wishlist/static/src/js/variant_mixin.js:4]: It seems to work without 🤔

File: addons/website_sale_stock/data/template_email.xml (hunk 14 above)
Comment (alpa-odoo) [reply addons/website_sale_stock/data/template_email.xml:19]: https://github.com/odoo/odoo/pull/87789#discussion_r884882621

File: addons/website_sale_stock_wishlist/static/src/js/website_sale.js (hunk 11 above)
Comment (Williambraecky) [reply addons/website_sale_stock_wishlist/static/src/js/website_sale.js:37]: Quite honestly it could be simpler, do we actually need the new record in this case? I don't think so..
Either way we can return something as simple as `{'error': message}` in case of error and `{'wish': wish}` in case of success.
The status code does not matter.

File: addons/website_sale_wishlist/static/src/js/website_sale.js [hunk 18]
```diff
@@ -13,4 +22,282 @@ WebsiteSale.include({
...
+        }).then(function (resp) {
+            if (resp.warning) {
+                if (!$('#data_warning').length) {
+                    $('.wishlist-section').prepend('<div class="mt16 alert alert-danger alert-dismissable" role="alert" id="data_warning"></div>');
+                }
+                let cart_alert = $('.wishlist-section').parent().find('#data_warning');
+                cart_alert.html('<button type="button" class="btn-close" data-bs-dismiss="alert" aria-hidden="true">&times;</button> ' + resp.warning);
+
+            }
+            $('.my_cart_quantity').html(resp.cart_quantity || '<i class="fa fa-warning" /> ');
```
Comment (Williambraecky) [inline addons/website_sale_wishlist/static/src/js/website_sale.js:199]: Those two `html` calls were moved from [here](https://github.com/odoo/odoo/blob/master/addons/website_sale_wishlist/static/src/js/website_sale_wishlist.js#L197-L200).

**Review by Williambraecky — COMMENTED** [review]
Review for @odoo/rd-security

File: addons/website_sale_stock_wishlist/static/src/xml/product_availability.xml [hunk 19]
```diff
@@ -2,29 +2,25 @@
...
-                type="button"
-                role="button"
-                class="btn btn-primary btn-lg text-nowrap my-2 o_add_wishlist_dyn"
-                t-att-data-product-template-id="product_id"
-                t-att-data-product-product-id="product_id"
-                data-action="o_wishlist"
-                title="Add to wishlist">
-                    <i class="fa fa-heart-o me-2" role="img" aria-label="Add to wishlist"/>
-                    Save for later
+                    id="save_for_later_button"
```
Comment (Williambraecky) [inline addons/website_sale_stock_wishlist/static/src/xml/product_availability.xml:8]: This should be named something more related to `website_sale` or wishlist.
```suggestion
                    id="wsale_save_for_later_button"
```

File: addons/website_sale_wishlist/controllers/main.py [hunk 20]
```diff
@@ -33,15 +33,14 @@ def add_to_wishlist(self, product_id, **kw):
             product_id,
             partner_id
         )
-
-        if not partner_id:
-            request.session['wishlist_ids'] = request.session.get('wishlist_ids', []) + [wish.id]
-
+        wish.displayed_in_cart = True
+        request.session['wishlist_ids'] = request.session.get('wishlist_ids', []) + [wish.id]
```
Comment (Williambraecky) [inline addons/website_sale_wishlist/controllers/main.py:37]:
```suggestion
        request.session.setdefault('wishlist_ids', []).append(wish.id)
```

File: addons/website_sale_wishlist/models/product_wishlist.py [hunk 21]
```diff
@@ -89,4 +49,4 @@ class ProductProduct(models.Model):
 
     def _is_in_wishlist(self):
         self.ensure_one()
-        return self in self.env['product.wishlist'].current().mapped('product_id')
+        return self in self.env['product.wishlist'].current().filtered(lambda wish: wish.displayed_in_cart).mapped('product_id')
```
Comment (Williambraecky) [inline addons/website_sale_wishlist/models/product_wishlist.py:52]:
```suggestion
        # displayed_in_cart is used to tell whether a wishlist record is in the actual wishlist or not.
        return self in self.env['product.wishlist'].current().filtered(lambda wish: wish.displayed_in_cart).mapped('product_id')
```
This comment could also be moved on the definition of `displayed_in_cart`

**Review by Williambraecky — CHANGES_REQUESTED** [review]
Thanks for your work!
Some small comments :)

File: addons/website_sale_wishlist/static/src/js/website_sale.js (hunk 18 above)
Comment (xmo-odoo) [reply addons/website_sale_wishlist/static/src/js/website_sale.js:199]: And the entire thing needs to be synchronised / removed, this feature was half-broken and got fixed in #95637 with the entire thing unified between `website_sale` and `website_sale_wishlist`, and the icon & menu updates moved to website_sale_utils.js.

File: addons/website_sale_wishlist/controllers/main.py [hunk 22]
```diff
@@ -59,5 +58,8 @@ def rm_from_wishlist(self, wish, **kw):
                 request.session.modified = True
                 wish.sudo().unlink()
         else:
-            wish.unlink()
+            if wish.stock_notification:
+                wish.displayed_in_cart = False
+            else:
+                wish.unlink()
```
Comment (xmo-odoo) [inline addons/website_sale_wishlist/controllers/main.py:64]:
```suggestion
        elif wish.stock_notification:
            wish.displayed_in_cart = False
        else:
            wish.unlink()
```

File: addons/website_sale/models/product_wishlist.py [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+    currency_id = fields.Many2one('res.currency', related='pricelist_id.currency_id', readonly=True)
+    pricelist_id = fields.Many2one('product.pricelist', string='Pricelist', help='Pricelist when added')
+    price = fields.Monetary(currency_field='currency_id', string='Price',
+                            help='Price of the product when it has been added in the wishlist')
+    website_id = fields.Many2one('website', ondelete='cascade', required=True)
+    active = fields.Boolean(default=True, required=True)
+    stock_notification = fields.Boolean(default=False, required=True)
+    displayed_in_cart = fields.Boolean(default=True, required=True,
+                                       help='Is used to tell whether a wishlist record should be displayed in the '
+                                            'user wishlist')
```
Comment (xmo-odoo) [inline addons/website_sale/models/product_wishlist.py:28]: If it's about being visible in the wishlist why is it called "visible_*in_cart*"?

Surely it should be `visible_in_wishlist`? Or given it's a wishlist item and there's nowhere else it *could* be displayed, why not just "visible" or "displayed"?

File: addons/website_sale_stock_wishlist/static/src/js/website_sale.js [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+    }),
+
+    _onClickAddStockEmailNotificationWishlistMessage: function (ev) {
+        //hide the message, display the input box
+        ev.currentTarget.classList.add('d-none');
+        const formEl = ev.currentTarget.parentElement.querySelector('#add_stock_email_notification_wishlist_form');
+        formEl.classList.remove('d-none');
+    },
+
+    _onSubmitAddStockEmailNotificationWishlistForm: function (ev) {
```
Comment (xmo-odoo) [inline addons/website_sale_stock_wishlist/static/src/js/website_sale.js:19]: This is brand new code, why is it still using long-form function attributes and callbacks-based futures?

File: addons/website_sale_wishlist/static/src/js/website_sale.js [hunk 25]
```diff
@@ -13,4 +22,282 @@ WebsiteSale.include({
...
+     * Gets the current wishlist items.
+     * In editable mode, do nothing instead.
+     *
+     * @override
+     */
+    willStart: function () {
+        let self = this;
+        let def = this._super.apply(this, arguments);
+
+        let wishDef = $.get('/shop/wishlist', {
```
Comment (xmo-odoo) [inline addons/website_sale_wishlist/static/src/js/website_sale.js:42]: The weird mix of modern javascript and old unnecessary jquery stuff (which isn't even really a gain) is rather strange.

**Review by xmo-odoo — COMMENTED** [review]
And stapling this feature to the wishlist seems to only add complexity to it, creating wishlist items which are not in the wishlist, and requiring additional flags to discriminate between wishlist wishlist items, notification wishlist items, and both.

This is made further nonsensical because it really has nothing to do with `website_sale`, or `website_sale_wishlist`, this does nothing without `website_sale_stock`.

Why is it not just `website_sale_stock` adding an m2m (or some more fleshed out but m2m-adjacent model if additional attributes are necessary) between partners and products?

File: addons/website_sale_wishlist/static/src/js/website_sale.js (hunk 18 above)
Comment (Williambraecky) [reply addons/website_sale_wishlist/static/src/js/website_sale.js:199]: Yes this code does not exist anymore :)

Comment (Williambraecky) [reply addons/website_sale_wishlist/static/src/js/website_sale.js:199]: Or at least it shouldn't.

**Comment by alpa-odoo** [comment]
> And stapling this feature to the wishlist seems to only add complexity to it, creating wishlist items which are not in the wishlist, and requiring additional flags to discriminate between wishlist wishlist items, notification wishlist items, and both.
> 
> This is made further nonsensical because it really has nothing to do with `website_sale`, or `website_sale_wishlist`, this does nothing without `website_sale_stock`.
> 
> Why is it not just `website_sale_stock` adding an m2m (or some more fleshed out but m2m-adjacent model if additional attributes are necessary) between partners and products?

Indeed thanks for the review.
